lib/Parse/ParseExpr.cpp

```diff
@@ -1009,6 +1009,10 @@ StringRef Parser::copyAndStripUnderscores(StringRef orig) {
   return StringRef(start, p - start);
 }
 
+StringRef Parser::copyAndStripUnderscoresIfNeeded(StringRef text) {
+  return text.contains('_') ? copyAndStripUnderscores(text) : text;
+}
```

Comment:
Sorry, this is confusing. The main purpose of `copyAndStripUnderscores()` is copying the literal string to `ASTContext` for later stages getting the literal value from it. Stripping underscores is a secondary "while you are here" thing.

However this new function is all about stripping underscores. We don't want to copy it to `ASTContext` if possible, and the copied string just "leaks." So I don't think we should use similar name for this.

If you want to factor out it to a function, could you do like:
```cpp
StringRef stripUnderscoresIfNeeded(StringRef str, SmallVectorImpl<char> &buf) {
  if (!str.contains('_')) return str;
  buf.clear();
  llvm::copy_if(str, std::back_inserter(buf), [](char c){ return c != '_'; });
  return StringRef(buffer.data(), buffer.size());
}
```
And use a temporary buffer like
```cpp
SmallString<16> scratch;
auto text = stripUnderscoresIfNeeded(Tok.getText(), scratch);
```
Otherwise, I don't think it's worth to make a function for this.